### PR TITLE
Bug fixes for test controller in add/remove switch and router

### LINKED ISSUE #29: Key to hash table for agent metadata flips between pointing to null and valid data.
**What happened**
When making consecutive update_agent_metadata calls, every other call will fail in the following scenario:
A droplet D1 is hosting endpoint e1 from VPC-1, and endpoint e2 from VPC-2.
e1 and e2 are both deleted
e1 and e2 are both recreated on the same droplet D1
Every other call to update_agent_metadata for e1 will fail
__________
Setting ep->data to null in the macro INTF_DELETE() seems to be the issue.
trn_transitd.h:81
However removing that line will reintroduce an old bug.

**What you expected to happen**
update_agent_metadata should not fail
**How to reproduce it**

    def setUp(self):

        self.droplets = {
            "d1": droplet("d1"),
            "d2": droplet("d2"),
            "d3": droplet("d3"),
            "d4": droplet("d4"),
            "switch-1": droplet("switch-1"),
            "switch-2": droplet("switch-2"),
            "router-1": droplet("router-1")
        }

        c = controller(self.droplets)

        # Create two co-hosted VPCs
        c.create_vpc(3, cidr("16", "10.0.0.0"), ["router-1"])
        c.create_network(3, 10, cidr("24", "10.0.0.0"), ["switch-1"])
        c.create_network(3, 20, cidr("24", "10.20.0.0"), ["switch-2"])

        self.ep1 = c.create_simple_endpoint(3, 10, "10.0.0.2", "d1")
        self.ep2 = c.create_simple_endpoint(3, 10, "10.0.0.3", "d2")

        self.ep3 = c.create_simple_endpoint(3, 20, "10.20.0.4", "d3")
        self.ep4 = c.create_simple_endpoint(3, 20, "10.20.0.5", "d4")

        c.create_vpc(4, cidr("16", "20.0.0.0"), ["router-1"])
        c.create_network(4, 10, cidr("24", "20.0.0.0"), ["switch-1"])
        c.create_network(4, 20, cidr("24", "20.20.0.0"), ["switch-2"])

        self.ep5 = c.create_simple_endpoint(4, 10, "20.0.0.2", "d1")
        self.ep6 = c.create_simple_endpoint(4, 10, "20.0.0.3", "d2")

        self.ep7 = c.create_simple_endpoint(4, 20, "20.20.0.4", "d3")
        self.ep8 = c.create_simple_endpoint(4, 20, "20.20.0.5", "d4")

        # Delete both VPCs
        c.delete_vpc(4, cidr("16", "10.0.0.0"))
        c.delete_vpc(3, cidr("16", "20.0.0.0"))

        # Recreate the two co-hosted VPCs
        c.create_vpc(3, cidr("16", "10.0.0.0"), ["router-1"])
        c.create_network(3, 10, cidr("24", "10.0.0.0"), ["switch-1"])
        c.create_network(3, 20, cidr("24", "10.20.0.0"), ["switch-2"])

        self.ep1 = c.create_simple_endpoint(3, 10, "10.0.0.2", "d1")
        self.ep2 = c.create_simple_endpoint(3, 10, "10.0.0.3", "d2")

        self.ep3 = c.create_simple_endpoint(3, 20, "10.20.0.4", "d3")
        self.ep4 = c.create_simple_endpoint(3, 20, "10.20.0.5", "d4")

        c.create_vpc(4, cidr("16", "20.0.0.0"), ["router-1"])
        c.create_network(4, 10, cidr("24", "20.0.0.0"), ["switch-1"])
        c.create_network(4, 20, cidr("24", "20.20.0.0"), ["switch-2"])

        self.ep5 = c.create_simple_endpoint(4, 10, "20.0.0.2", "d1")
        self.ep6 = c.create_simple_endpoint(4, 10, "20.0.0.3", "d2")

        self.ep7 = c.create_simple_endpoint(4, 20, "20.20.0.4", "d3")
        self.ep8 = c.create_simple_endpoint(4, 20, "20.20.0.5", "d4")
        while(True):
            sleep(100000)

**Environment**
Ubuntu Server 18.04 LTS (HVM) on ec2 t2.2xlarge

### PR DESCRIPTION
**Pull request by phudtran**
test: Fixes some bugs in the test controller.

This PR does the following:
Fixes bug in add/remove switch and router.
Fixes bug when reclaiming veth from phantom gateway.
Adds another functional test to validate remove switch/router functionality.

### CODE REVIEW

File: src/dmn/trn_transitd.h [hunk 1]
```diff
@@ -69,7 +69,6 @@
 		e.key = itf;                                                   \
 		ep = hsearch(e, FIND);                                         \
 		if (ep) {                                                      \
-			free(ep->key);                                         \
```
Comment (zasherif) [inline src/dmn/trn_transitd.h]: I don't fully understand why this causes the table to return a value that alternate between ep->data == NULL and ep->data is not NULL. What am I missing?

File: test/trn_func_tests/test_delete_vpc_change_switch_router.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        self.c.delete_vpc(3, cidr("16", "10.0.0.0"))
+
+    def tearDown(self):
+        pass
+
+    def test_delete_vpc_change_switch_router(self):
+        logger.info(
+            "{} Testing delete vpc with added/removed routers and switches! {}".format('='*20, '='*20))
+        # do_validate_delete_test(self, list(self.droplets.values()))
```
Comment (zasherif) [inline test/trn_func_tests/test_delete_vpc_change_switch_router.py]: Why this line is commented?

File: test/trn_func_tests/test_delete_vpc_change_switch_router.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        self.c.add_router(3, "router-2")
+        self.c.add_switch(3, 10, "d3")
+        self.c.add_switch(3, 10, "d5")
+        self.c.add_switch(3, 20, "d2")
+        self.c.add_switch(3, 20, "d5")
+        self.c.add_router(3, "d5")
+        self.c.remove_switch(3, 20, "d2")
+        self.c.remove_router(3, "d5")
+
+        self.c.delete_vpc(3, cidr("16", "10.0.0.0"))
```
Comment (zasherif) [inline test/trn_func_tests/test_delete_vpc_change_switch_router.py]: The VPC id is unique in the test, do you need to pass the cidr as well?

File: test/trn_func_tests/test_delete_one_at_a_time.py [hunk 4]
```diff
@@ -94,8 +94,8 @@ def setUp(self):
 
         # Delete an endpoint in subnet, delete the parent subnet, then parent VPC.
         self.ep3 = c.delete_simple_endpoint(3, 20, "10.20.0.4", "d3")
-        c.delete_network(3, 20, cidr("24", "10.20.0.0"), ["switch-2"])
-        c.delete_vpc(3, cidr("16", "10.0.0.0"), ["router-1"])
+        c.delete_network(3, 20, cidr("24", "10.20.0.0"))
```
Comment (zasherif) [inline test/trn_func_tests/test_delete_one_at_a_time.py]: Why do you need the cidr to call delete_network?

File: test/trn_func_tests/test_delete_vpc_change_switch_router.py (hunk 3 above)
Comment (phudtran) [reply test/trn_func_tests/test_delete_vpc_change_switch_router.py]: I don't. Fixing now.

File: test/trn_func_tests/test_delete_one_at_a_time.py (hunk 4 above)
Comment (phudtran) [reply test/trn_func_tests/test_delete_one_at_a_time.py]: I don't. Fixing now.

**Pull request merged**